### PR TITLE
 Re : no attribute 'lookahead'

### LINKED ISSUE #11352: coset_enumeration_c raises "AttributeError: 'CosetTable' object has no attribute 'lookahead'"
```
gaurav@stallman:~/Public/sympy$ python3
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
[GCC 4.8.4] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> from sympy.combinatorics.free_group import free_group
>>> from sympy.combinatorics.fp_groups import FpGroup, coset_enumeration_c, coset_enumeration_r
>>> F, a, b, c = free_group("a, b, c")
>>> f = FpGroup(F, [a**11, b**5, c**4, (a*c)**3, b**2*c**-1*b**-1*c, a**4*b**-1*a**-1*b])
>>> H = [c, b, c**2]
>>> C = coset_enumeration_c(f, H)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/gaurav/Public/sympy/sympy/combinatorics/fp_groups.py", line 793, in coset_enumeration_c
    C.process_deductions(R_c_list[C.A_dict[x]], R_c_list[C.A_dict_inv[x]])
  File "/home/gaurav/Public/sympy/sympy/combinatorics/fp_groups.py", line 466, in process_deductions
    self.lookahead()
AttributeError: 'CosetTable' object has no attribute 'lookahead'
```

### PR DESCRIPTION
**Pull request by kritkaran94**
Fixes #11352

### CODE REVIEW

**Comment by kritkaran94** [comment]
@gxyd Please review this PR.

**Comment by kritkaran94** [comment]
@gxyd Does this Look good ?

File: sympy/combinatorics/tests/test_fp_groups.py [hunk 1]
```diff
@@ -125,6 +125,413 @@ def test_coset_enumeration():
     # this test function contains the combined tests for the two strategies
     # i.e. HLT and Felsch strategies.
 
+    # Test for Issue #11352
+    F, a, b, c = free_group("a, b, c")
+    f = FpGroup(F, [a**11, b**5, c**4, (a*c)**3, b**2*c**-1*b**-1*c, a**4*b**-1*a**-1*b])
+    H = [c, b, c**2]
+    C_r = coset_enumeration_c(f, H)
+    C_r.compress(); C_r.standardize()
```
Comment (gxyd) [inline sympy/combinatorics/tests/test_fp_groups.py]: Remove this line. And the line above this i.e remove lines [132](https://github.com/sympy/sympy/pull/11356/files#diff-07f5c9eafa6f27616138278a54dd0e75R132) and [133](https://github.com/sympy/sympy/pull/11356/files#diff-07f5c9eafa6f27616138278a54dd0e75R133) . Sine testing is to be there only for `coset_enumeration_c`.

File: sympy/combinatorics/tests/test_fp_groups.py [hunk 2]
```diff
@@ -125,6 +125,413 @@ def test_coset_enumeration():
...
+             [276, 256, 254, 358, 261, 390],
+             [344, 334, 261, 392, 329, 359],
+             [335, 262, 355, 305, 344, 355],
+             [264, 288, 339, 360, 387, 328],
+             [352, 275, 277, 330, 364, 363],
+             [363, 309, 388, 319, 275, 358],
+             [321, 306, 350, 394, 284, 283],
+             [291, 349, 393, 385, 383, 382],
+             [376, 314, 322, 378, 365, 370]]
+    assert C_r.table == table0
```
Comment (gxyd) [inline sympy/combinatorics/tests/test_fp_groups.py]: Change `C_r.table` -> `C_c.table`

File: sympy/combinatorics/tests/test_fp_groups.py [hunk 3]
```diff
@@ -125,6 +125,413 @@ def test_coset_enumeration():
...
+    assert C_c.table == table0
```
Comment (gxyd) [inline sympy/combinatorics/tests/test_fp_groups.py]: Remove this extra line.

File: sympy/combinatorics/tests/test_fp_groups.py (hunk 1 above)
Comment (kritkaran94) [inline sympy/combinatorics/tests/test_fp_groups.py]: @gxyd You want to me remove C_r = coset_enumeration_c(f,H) and the following line too ?

Comment (gxyd) [inline sympy/combinatorics/tests/test_fp_groups.py]: Yes, that will be good.

**Comment by kritkaran94** [comment]
@gxyd Fixed those.

**Comment by gxyd** [comment]
This looks fine. Will merge in 24 hrs.

**Comment by kritkaran94** [comment]
@gxyd Thanks for that.

**Comment by jksuom** [comment]
Would it be possible to test `look_ahead` by lowering `max_stack_size` for the test to avoid the big hard coded table? One could first construct `table0` (of smaller size) with the standard setting, and then change the limit and compare the results.

**Comment by kritkaran94** [comment]
@jksuom We could plan to have a smaller test for sure. This particular test was included as this was the example mentioned in the issue created by @gxyd. 

We could change the test. Do we need to replace the test ? 

I feel having one more long hard coded table isn't that bad as we already have large tables in the test file.

**Comment by jksuom** [comment]
I think tests that produce the result in two different ways are more reliable than those comparing a single result with precomputed data. When that data is produced by the same function, a comparison does not suffice to prove its correctness.

**Comment by gxyd** [comment]
I think that would be possible. But I hope you don't expect this to 
reduce the "time of this test" in any way (since I have some other thing 
in mind as well, just answering this much question would suffice).

**Comment by gxyd** [comment]
I meant the issue that remains to be addressed is "space occupying" by hard coding of "table", not in terms of any issue of time taken by test.

**Comment by kritkaran94** [comment]
@gxyd So, I feel that we should stick with the tests we have at present.

**Comment by gxyd** [comment]
No, I think we should change it. @jksuom is probably right. Also there 
are a few other things that will get fixed by modifying the tests.

**Comment by kritkaran94** [comment]
@gxyd Okay, so what should changes should I make in this PR ?

**Comment by gxyd** [comment]
I have sent in a PR https://github.com/kritkaran94/sympy/pull/6 w.r.t your branch `patch-1` (I have done it this way for first time, not sure if any issues are there), it will probably fix the issues in tests. Please merge that PR into your branch **only after it passes the tests**.

**Comment by kritkaran94** [comment]
@gxyd, Sure I would merge the PR to my branch as soon as the tests pass. Thanks a lot for that.

**Comment by gxyd** [comment]
Also please wait for @jksuom give +1 on that.

**Comment by kritkaran94** [comment]
@gxyd For sure :)

**Comment by kritkaran94** [comment]
@gxyd I merged your PR as all the tests passed. @jksuom Can you please review this PR?

**Comment by jksuom** [comment]
I think this looks good. +1 to merge when the tests pass.

**Comment by kritkaran94** [comment]
@jksuom I can't see the tests running on travis. The tests did pass [locally.](https://github.com/kritkaran94/sympy/pull/6) Any idea on this ?

**Comment by jksuom** [comment]
It seems that there are also merge conflicts. If you take care of those and push a new commit, travis will probably start working.

**Comment by gxyd** [comment]
I think that would be possible. But I hope you don't expect this to reduce
the "time of this test" in any way.

On 07/09/2016 12:37 PM, Kalevi Suominen wrote:

Would it be possible to test look_ahead by lowering max_stack_size for the
test to avoid the big hard coded table? One could first construct table0
(of smaller size) with the standard setting, and then change the limit and
compare the results.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/sympy/sympy/pull/11356#issuecomment-231520035>, or mute
the thread
<https://github.com/notifications/unsubscribe/AHWt8dN6mCO_WK9iaE55gMve2-uj9LuJks5qT0i_gaJpZM4JHX4Q>
.

**Comment by gxyd** [comment]
Please ignore this last message. It was kept as a draft in my mail box and I mistakenly pressed send.

**Comment by kritkaran94** [comment]
@jksuom Added a new commit too with a minor change. I guess the tests are still not working. What can we do now on this ?

@gxyd any idea on this ?

**Comment by gxyd** [comment]
You can follow the [github article on merging when conflicts](https://help.github.com/articles/resolving-a-merge-conflict-from-the-command-line/) are there. If you **still** have issues with that (as a final option), you can merge the PR https://github.com/kritkaran94/sympy/pull/7 , that will fix it.

**Comment by kritkaran94** [comment]
@gxyd Thanks a lot for the information. I will merge the PR and drop the "WIP" too from the title.

**Comment by kritkaran94** [comment]
@jksuom, @gxyd All the tests have passed. Can this be merged now ?

**Comment by jksuom** [comment]
Yes, I think it is good now. Thanks!

**Pull request merged**